### DISCUSSION TITLE
LED blink error - cannot join nonexistent thread

### DISCUSSION THREAD

**Discussion by westdale-design** [gpiozero/gpiozero#1182] [Q&A]
I may be missing something here - advice welcome - possibly a bypass to test whether a certain thread exists? 
Under some circumstances (I have a program that has a power-fail reboot so not all runs are the same) ... the start of a blink on an LED fails because the first action of `blink(self, on_time=1, off_time=1, n=None, background=True)`:    is `self._stop_blink()` on that LED and that fails in trying to join the blink thread if there isn't one.

```python
  File "/home/pi/cellarmonitor.py", line 207, in run
    set_leds()
  File "/home/pi/cellarmonitor.py", line 119, in set_leds
    red_led.blink(on_time = 0.5, off_time=0.5)
  File "/usr/lib/python3/dist-packages/gpiozero/output_devices.py", line 236, in blink
    self._stop_blink()
  File "/usr/lib/python3/dist-packages/gpiozero/output_devices.py", line 249, in _stop_blink
    self._blink_thread.stop()
  File "/usr/lib/python3/dist-packages/gpiozero/threads.py", line 46, in stop
    self.join(timeout)
  File "/usr/lib/python3/dist-packages/gpiozero/threads.py", line 49, in join
    super().join(timeout)
  File "/usr/lib/python3.11/threading.py", line 1107, in join
    raise RuntimeError("cannot join thread before it is started")
RuntimeError: cannot join thread before it is started
```

**Comment by lurch** [comment]
```
  File "/usr/lib/python3/dist-packages/gpiozero/output_devices.py", line 249, in _stop_blink
    self._blink_thread.stop()
```
Hmmm, the `_stop_blink` method does:
```python
    def _stop_blink(self):
        if getattr(self, '_controller', None):
            self._controller._stop_blink(self)
        self._controller = None
        if getattr(self, '_blink_thread', None):
            self._blink_thread.stop()
        self._blink_thread = None
```
So somehow `self._blink_thread` is **not** `None`, but the thread hasn't actually started yet?

The `__init__` method does `self._blink_thread = None` and then the `blink` method does:
```python
    def blink(
            self, on_time=1, off_time=1, fade_in_time=0, fade_out_time=0,
            n=None, background=True):
        self._stop_blink()
        self._blink_thread = GPIOThread(
            self._blink_device,
            (on_time, off_time, fade_in_time, fade_out_time, n)
        )
        self._blink_thread.start()
        if not background:
            self._blink_thread.join()
            self._blink_thread = None
```
which implies that you're somehow managing to call `_stop_blink` _after_ `self._blink_thread = GPIOThread(self._blink_device, (on_time, off_time, fade_in_time, fade_out_time, n))` but _before_ `self._blink_thread.start()`, which doesn't seem possible.
Hmmm, is it possible that your Pi is running out of memory, and so whilst Python is able to create the new thread, it isn't able to start it?

**Reply by westdale-design** [reply]
Thank you for your detailed thoughts.   Although the basic program on this pizero is trivial - receive tcp broadcast event from the pump program (a pizero2) and show appropriate LEDs for low-level, pump-running etc  - there is quite a bit of  surrounding code: the tcp handler and  a telegram robot (using 'requests') for alarms, 2 threads to check pump has been sending/receiving messages recently -  and it records events in a circular buffer for diagnosis. 
I was surprised at the error as the code (below) to handle the LEDs turns them all off before trying to start the blink or other states. I'll also look at the difference between a full reboot and a rerun.... I assumed that all blink threads , like my own monitoring threads, would be daemons and terminate with the program so a rerun for testing would be clean for testing as if at reboot (at least in this respect ) . 
One thing I will definitely try is reducing the size of the event buffer.  Also you trigger a thought that the pizero may be too slow even though the events are of the order of 10 seconds - and the garbage collector may not be effective enough. Perhaps my simplest solution may turn out to be another pizero2 :-)

 # routines to set leds and state_text  - echo state on pump
def leds_off(_all=False):
    green_led.off()
...
    if _all: blue_led.off()
    # blue led left alight as power indicator - but can clear at startup

def set_leds():
    # set the LEDs according to the latest event information 
    e = Cellar_state.latest_event()
    leds_off()
    # Pump state
    if Cellar_state.pump_down:
        # simple warning - there will be a telegram sent as well
        white_led.blink(on_time=0.5, off_time=0.5)
        green_led.blink(on_time=0.5, off_time=0.5)
    elif e.pump:
...
    if e.flood:
       red_led.blink(on_time = 0.5, off_time=0.5)   <<<<<<<<<<<
       yellow_led.blink(on_time = 0.5, off_time=0.5)
...
    else:
        # show values of the low and high probes
        if e.low: yellow_led.on()
        if e.high: red_led.on()
